internal/collect_es6_sources.bzl

```diff
@@ -0,0 +1,57 @@
+# Copyright 2017 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Used by production rules to expose a file tree of only es6 files.
+
+These are expected to be used by the production toolchain, such as bundlers.
+The tree will be flattened, such that all the es6 files are under a single tree.
+"""
+
+def collect_es6_sources(ctx):
+  """Returns a file tree containing only production files.
+
+  Args:
+    ctx: ctx.
+
+  Returns:
+    A file tree containing only production files.
+  """
+
+  non_rerooted_files = depset()
+  for dep in ctx.attr.deps:
+    if hasattr(dep, "typescript"):
+      non_rerooted_files = depset(transitive = [non_rerooted_files, dep.typescript.transitive_es6_sources])
+    else:
+      fail(
+          ("%s is neither a TypeScript nor a Closure JS library producing rule." % dep.label) +
+          "\nDependencies must be ts_library, ts_declaration, or closure_js_library.")
+
+  rerooted_files = []
+  for file in non_rerooted_files.to_list():
+    rerooted_file = ctx.actions.declare_file(
+      "%s.closure/%s" % (
+        ctx.label.name,
+        file.short_path.replace(".closure.js", ".js")))
```

Comment:
Leave a comment to update this to only target the end of the short path (there's potential for a stupidly named file being sad)